File: sdk/nodejs/src/delegation.ts
/**
 * SatGate Delegation Helpers
 * 
 * Utilities for creating and managing macaroon delegation hierarchies.
 * Supports common patterns like team budgets, time-limited access, and scope restrictions.
 */

import { SatGateClient } from './client';
import { Token, DelegateRequest } from './types';

/**
 * Caveat builders for common delegation patterns
 */
export const Caveats = {
  /**
   * Restrict scope to specific actions
   * @example Caveats.scope('api:read')
   */
  scope: (scope: string): string => `scope = ${scope}`,

  /**
   * Set expiration time
   * @param seconds Seconds from now, or Date object
   */
  expires: (seconds: number | Date): string => {
    const expiry = typeof seconds === 'number'
      ? Math.floor(Date.now() / 1000) + seconds
      : Math.floor(seconds.getTime() / 1000);
    return `expires = ${expiry}`;
  },

  /**
   * Limit to specific routes
   * @example Caveats.routes(['/api/v1/*', '/health'])
   */
  routes: (patterns: string[]): string => `routes = ${patterns.join(',')}`,

  /**
   * Limit requests per period
   * @param count Max requests
   * @param period Period in seconds (default: 60 = per minute)
   */
  rateLimit: (count: number, period = 60): string => 
    `ratelimit = ${count}/${period}`,

  /**
   * Set a budget limit (for billing/metering)
   * @param amount Budget amount
   * @param currency Currency code (default: 'USD')
   */
  budget: (amount: number, currency = 'USD'): string =>
    `budget = ${amount} ${currency}`,

  /**
   * Restrict to specific IP addresses/CIDRs
   * @example Caveats.sourceIp(['10.0.0.0/8', '192.168.1.100'])
   */
  sourceIp: (cidrs: string[]): string => `source_ip = ${cidrs.join(',')}`,

  /**
   * Restrict to specific methods
   * @example Caveats.methods(['GET', 'POST'])
   */
  methods: (methods: string[]): string => `methods = ${methods.join(',')}`,

  /**
   * Add a team/cost-center identifier for attribution
   */
  team: (teamId: string): string => `team = ${teamId}`,

  /**
   * Add a project identifier for attribution
   */
  project: (projectId: string): string => `project = ${projectId}`,

  /**
   * Add a custom label for tracking
   */
  label: (key: string, value: string): string => `label:${key} = ${value}`,

  /**
   * Add an arbitrary caveat
   */
  custom: (caveat: string): string => caveat,
};

/**
 * DelegationBuilder provides a fluent API for creating delegated tokens
 */
export class DelegationBuilder {
  private parentToken: string;
  private caveats: string[] = [];

  constructor(parentToken: string) {
    this.parentToken = parentToken;
  }

  /** Add a scope restriction */
  withScope(scope: string): this {
    this.caveats.push(Caveats.scope(scope));
    return this;
  }

  /** Add an expiration */
  withExpiry(seconds: number | Date): this {
    this.caveats.push(Caveats.expires(seconds));
    return this;
  }

  /** Add route restrictions */
  withRoutes(patterns: string[]): this {
    this.caveats.push(Caveats.routes(patterns));
    return this;
  }

  /** Add rate limiting */
  withRateLimit(count: number, periodSeconds = 60): this {
    this.caveats.push(Caveats.rateLimit(count, periodSeconds));
    return this;
  }

  /** Add a budget limit */
  withBudget(amount: number, currency = 'USD'): this {
    this.caveats.push(Caveats.budget(amount, currency));
    return this;
  }

  /** Restrict to source IPs */
  withSourceIp(cidrs: string[]): this {
    this.caveats.push(Caveats.sourceIp(cidrs));
    return this;
  }

  /** Restrict to HTTP methods */
  withMethods(methods: string[]): this {
    this.caveats.push(Caveats.methods(methods));
    return this;
  }

  /** Add team attribution */
  forTeam(teamId: string): this {
    this.caveats.push(Caveats.team(teamId));
    return this;
  }

  /** Add project attribution */
  forProject(projectId: string): this {
    this.caveats.push(Caveats.project(projectId));
    return this;
  }

  /** Add a custom label */
  withLabel(key: string, value: string): this {
    this.caveats.push(Caveats.label(key, value));
    return this;
  }

  /** Add a raw caveat */
  withCaveat(caveat: string): this {
    this.caveats.push(caveat);
    return this;
  }

  /** Build the delegation request */
  build(): DelegateRequest {
    return {
      parentToken: this.parentToken,
      caveats: this.caveats,
    };
  }

  /** Execute the delegation via client */
  async delegate(client: SatGateClient): Promise<Token> {
    return client.tokens.delegate(this.build());
  }
}

/**
 * Create a new delegation builder
 * @param parentToken The parent token to delegate from
 */
export function delegate(parentToken: string): DelegationBuilder {
  return new DelegationBuilder(parentToken);
}

/**
 * Common delegation patterns
 */
export const DelegationPatterns = {
  /**
   * Create a read-only token
   */
  readOnly: (parentToken: string): DelegationBuilder =>
    delegate(parentToken)
      .withScope('api:read')
      .withMethods(['GET', 'HEAD', 'OPTIONS']),

  /**
   * Create a time-limited token (24 hours by default)
   */
  temporary: (parentToken: string, hours = 24): DelegationBuilder =>
    delegate(parentToken)
      .withExpiry(hours * 3600),

  /**
   * Create a team token with budget
   */
  teamBudget: (
    parentToken: string,
    teamId: string,
    budgetAmount: number,
    currency = 'USD'
  ): DelegationBuilder =>
    delegate(parentToken)
      .forTeam(teamId)
      .withBudget(budgetAmount, currency),

  /**
   * Create a rate-limited API client token
   */
  apiClient: (
    parentToken: string,
    clientId: string,
    requestsPerMinute = 100
  ): DelegationBuilder =>
    delegate(parentToken)
      .withLabel('client_id', clientId)
      .withRateLimit(requestsPerMinute, 60),

  /**
   * Create a webhook callback token (single route, short expiry)
   */
  webhook: (
    parentToken: string,
    callbackPath: string,
    expiryMinutes = 30
  ): DelegationBuilder =>
    delegate(parentToken)
      .withRoutes([callbackPath])
      .withMethods(['POST'])
      .withExpiry(expiryMinutes * 60),

  /**
   * Create a CI/CD pipeline token (specific routes, short-lived)
   */
  cicd: (
    parentToken: string,
    pipelineId: string,
    expiryMinutes = 60
  ): DelegationBuilder =>
    delegate(parentToken)
      .withLabel('pipeline', pipelineId)
      .withExpiry(expiryMinutes * 60),

  /**
   * Create an agent swarm token (budget + rate limit)
   */
  agentSwarm: (
    parentToken: string,
    swarmId: string,
    options: {
      budget?: number;
      currency?: string;
      requestsPerMinute?: number;
      maxAgents?: number;
    } = {}
  ): DelegationBuilder => {
    let builder = delegate(parentToken)
      .withLabel('swarm_id', swarmId);
    
    if (options.budget) {
      builder = builder.withBudget(options.budget, options.currency || 'USD');
    }
    if (options.requestsPerMinute) {
      builder = builder.withRateLimit(options.requestsPerMinute, 60);
    }
    if (options.maxAgents) {
      builder = builder.withLabel('max_agents', String(options.maxAgents));
    }
    
    return builder;
  },
};

/**
 * Delegation tree node for visualizing delegation hierarchies
 */
export interface DelegationNode {
  signature: string;
  scope?: string;
  team?: string;
  project?: string;
  budget?: string;
  expiresAt?: Date;
  children: DelegationNode[];
  metadata: Record<string, string>;
}

/**
 * Parse caveats from a token into structured data
 */
export function parseCaveats(caveats: string[]): Record<string, string> {
  const result: Record<string, string> = {};
  
  for (const caveat of caveats) {
    const match = caveat.match(/^([^=]+)\s*=\s*(.+)$/);
    if (match) {
      result[match[1].trim()] = match[2].trim();
    }
  }
  
  return result;
}

/**
 * Check if a caveat would be more restrictive than the parent
 */
export function isMoreRestrictive(
  parentCaveats: string[],
  childCaveat: string
): boolean {
  const [key] = childCaveat.split('=').map(s => s.trim());
  const parentParsed = parseCaveats(parentCaveats);
  const childParsed = parseCaveats([childCaveat]);
  
  // If parent doesn't have this caveat, child is more restrictive
  if (!(key in parentParsed)) {
    return true;
  }
  
  // For numeric caveats, child must be <= parent
  const numericCaveats = ['budget', 'ratelimit', 'expires'];
  if (numericCaveats.some(nc => key.includes(nc))) {
    const parentVal = parseFloat(parentParsed[key]) || Infinity;
    const childVal = parseFloat(childParsed[key]) || 0;
    return childVal <= parentVal;
  }
  
  // For scope caveats, child must be subset of parent
  if (key === 'scope') {
    const parentScopes = parentParsed[key].split(',');
    const childScopes = childParsed[key].split(',');
    return childScopes.every(cs => 
      parentScopes.some(ps => 
        ps === cs || ps === '*' || (ps.endsWith(':*') && cs.startsWith(ps.slice(0, -1)))
      )
    );
  }
  
  return true; // Default to allowing
}

File: sdk/python/satgate/delegation.py
"""
SatGate Delegation Helpers

Utilities for creating and managing macaroon delegation hierarchies.
Supports common patterns like team budgets, time-limited access, and scope restrictions.
"""

from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import List, Dict, Optional, Any, TYPE_CHECKING
import re

if TYPE_CHECKING:
    from .client import SatGateClient
    from .models import Token


class Caveats:
    """Caveat builders for common delegation patterns"""
    
    @staticmethod
    def scope(scope: str) -> str:
        """
        Restrict scope to specific actions
        
        Example:
            Caveats.scope('api:read')
        """
        return f"scope = {scope}"
    
    @staticmethod
    def expires(seconds: int = None, dt: datetime = None) -> str:
        """
        Set expiration time
        
        Args:
            seconds: Seconds from now
            dt: Specific datetime
        """
        if dt:
            expiry = int(dt.timestamp())
        elif seconds:
            expiry = int(datetime.now().timestamp()) + seconds
        else:
            raise ValueError("Either 'seconds' or 'dt' must be provided")
        return f"expires = {expiry}"
    
    @staticmethod
    def routes(patterns: List[str]) -> str:
        """
        Limit to specific routes
        
        Example:
            Caveats.routes(['/api/v1/*', '/health'])
        """
        return f"routes = {','.join(patterns)}"
    
    @staticmethod
    def rate_limit(count: int, period: int = 60) -> str:
        """
        Limit requests per period
        
        Args:
            count: Max requests
            period: Period in seconds (default: 60 = per minute)
        """
        return f"ratelimit = {count}/{period}"
    
    @staticmethod
    def budget(amount: float, currency: str = "USD") -> str:
        """
        Set a budget limit (for billing/metering)
        
        Args:
            amount: Budget amount
            currency: Currency code (default: 'USD')
        """
        return f"budget = {amount} {currency}"
    
    @staticmethod
    def source_ip(cidrs: List[str]) -> str:
        """
        Restrict to specific IP addresses/CIDRs
        
        Example:
            Caveats.source_ip(['10.0.0.0/8', '192.168.1.100'])
        """
        return f"source_ip = {','.join(cidrs)}"
    
    @staticmethod
    def methods(methods: List[str]) -> str:
        """
        Restrict to specific methods
        
        Example:
            Caveats.methods(['GET', 'POST'])
        """
        return f"methods = {','.join(methods)}"
    
    @staticmethod
    def team(team_id: str) -> str:
        """Add a team/cost-center identifier for attribution"""
        return f"team = {team_id}"
    
    @staticmethod
    def project(project_id: str) -> str:
        """Add a project identifier for attribution"""
        return f"project = {project_id}"
    
    @staticmethod
    def label(key: str, value: str) -> str:
        """Add a custom label for tracking"""
        return f"label:{key} = {value}"
    
    @staticmethod
    def custom(caveat: str) -> str:
        """Add an arbitrary caveat"""
        return caveat


class DelegationBuilder:
    """
    Fluent builder for creating delegated tokens
    
    Example:
        token = (DelegationBuilder(parent_token)
            .with_scope('api:read')
            .with_expiry(3600)
            .for_team('engineering')
            .delegate(client))
    """
    
    def __init__(self, parent_token: str):
        self._parent_token = parent_token
        self._caveats: List[str] = []
    
    def with_scope(self, scope: str) -> 'DelegationBuilder':
        """Add a scope restriction"""
        self._caveats.append(Caveats.scope(scope))
        return self
    
    def with_expiry(self, seconds: int = None, dt: datetime = None) -> 'DelegationBuilder':
        """Add an expiration"""
        self._caveats.append(Caveats.expires(seconds=seconds, dt=dt))
        return self
    
    def with_routes(self, patterns: List[str]) -> 'DelegationBuilder':
        """Add route restrictions"""
        self._caveats.append(Caveats.routes(patterns))
        return self
    
    def with_rate_limit(self, count: int, period_seconds: int = 60) -> 'DelegationBuilder':
        """Add rate limiting"""
        self._caveats.append(Caveats.rate_limit(count, period_seconds))
        return self
    
    def with_budget(self, amount: float, currency: str = "USD") -> 'DelegationBuilder':
        """Add a budget limit"""
        self._caveats.append(Caveats.budget(amount, currency))
        return self
    
    def with_source_ip(self, cidrs: List[str]) -> 'DelegationBuilder':
        """Restrict to source IPs"""
        self._caveats.append(Caveats.source_ip(cidrs))
        return self
    
    def with_methods(self, methods: List[str]) -> 'DelegationBuilder':
        """Restrict to HTTP methods"""
        self._caveats.append(Caveats.methods(methods))
        return self
    
    def for_team(self, team_id: str) -> 'DelegationBuilder':
        """Add team attribution"""
        self._caveats.append(Caveats.team(team_id))
        return self
    
    def for_project(self, project_id: str) -> 'DelegationBuilder':
        """Add project attribution"""
        self._caveats.append(Caveats.project(project_id))
        return self
    
    def with_label(self, key: str, value: str) -> 'DelegationBuilder':
        """Add a custom label"""
        self._caveats.append(Caveats.label(key, value))
        return self
    
    def with_caveat(self, caveat: str) -> 'DelegationBuilder':
        """Add a raw caveat"""
        self._caveats.append(caveat)
        return self
    
    def build(self) -> Dict[str, Any]:
        """Build the delegation request dictionary"""
        return {
            "parentToken": self._parent_token,
            "caveats": self._caveats,
        }
    
    def delegate(self, client: 'SatGateClient') -> 'Token':
        """Execute the delegation via client"""
        return client.tokens.delegate(
            parent_token=self._parent_token,
            caveats=self._caveats
        )


def delegate(parent_token: str) -> DelegationBuilder:
    """
    Create a new delegation builder
    
    Args:
        parent_token: The parent token to delegate from
    
    Example:
        token = delegate(parent_token).with_scope('api:read').delegate(client)
    """
    return DelegationBuilder(parent_token)


class DelegationPatterns:
    """Common delegation patterns"""
    
    @staticmethod
    def read_only(parent_token: str) -> DelegationBuilder:
        """Create a read-only token"""
        return (delegate(parent_token)
            .with_scope('api:read')
            .with_methods(['GET', 'HEAD', 'OPTIONS']))
    
    @staticmethod
    def temporary(parent_token: str, hours: int = 24) -> DelegationBuilder:
        """Create a time-limited token (24 hours by default)"""
        return delegate(parent_token).with_expiry(seconds=hours * 3600)
    
    @staticmethod
    def team_budget(
        parent_token: str,
        team_id: str,
        budget_amount: float,
        currency: str = "USD"
    ) -> DelegationBuilder:
        """Create a team token with budget"""
        return (delegate(parent_token)
            .for_team(team_id)
            .with_budget(budget_amount, currency))
    
    @staticmethod
    def api_client(
        parent_token: str,
        client_id: str,
        requests_per_minute: int = 100
    ) -> DelegationBuilder:
        """Create a rate-limited API client token"""
        return (delegate(parent_token)
            .with_label('client_id', client_id)
            .with_rate_limit(requests_per_minute, 60))
    
    @staticmethod
    def webhook(
        parent_token: str,
        callback_path: str,
        expiry_minutes: int = 30
    ) -> DelegationBuilder:
        """Create a webhook callback token (single route, short expiry)"""
        return (delegate(parent_token)
            .with_routes([callback_path])
            .with_methods(['POST'])
            .with_expiry(seconds=expiry_minutes * 60))
    
    @staticmethod
    def cicd(
        parent_token: str,
        pipeline_id: str,
        expiry_minutes: int = 60
    ) -> DelegationBuilder:
        """Create a CI/CD pipeline token (specific routes, short-lived)"""
        return (delegate(parent_token)
            .with_label('pipeline', pipeline_id)
            .with_expiry(seconds=expiry_minutes * 60))
    
    @staticmethod
    def agent_swarm(
        parent_token: str,
        swarm_id: str,
        budget: float = None,
        currency: str = "USD",
        requests_per_minute: int = None,
        max_agents: int = None
    ) -> DelegationBuilder:
        """Create an agent swarm token (budget + rate limit)"""
        builder = (delegate(parent_token)
            .with_label('swarm_id', swarm_id))
        
        if budget is not None:
            builder = builder.with_budget(budget, currency)
        if requests_per_minute is not None:
            builder = builder.with_rate_limit(requests_per_minute, 60)
        if max_agents is not None:
            builder = builder.with_label('max_agents', str(max_agents))
        
        return builder


@dataclass
class DelegationNode:
    """Delegation tree node for visualizing delegation hierarchies"""
    signature: str
    scope: Optional[str] = None
    team: Optional[str] = None
    project: Optional[str] = None
    budget: Optional[str] = None
    expires_at: Optional[datetime] = None
    children: List['DelegationNode'] = field(default_factory=list)
    metadata: Dict[str, str] = field(default_factory=dict)


def parse_caveats(caveats: List[str]) -> Dict[str, str]:
    """Parse caveats from a token into structured data"""
    result = {}
    for caveat in caveats:
        match = re.match(r'^([^=]+)\s*=\s*(.+)$', caveat)
        if match:
            result[match.group(1).strip()] = match.group(2).strip()
    return result


def is_more_restrictive(
    parent_caveats: List[str],
    child_caveat: str
) -> bool:
    """Check if a caveat would be more restrictive than the parent"""
    key = child_caveat.split('=')[0].strip()
    parent_parsed = parse_caveats(parent_caveats)
    child_parsed = parse_caveats([child_caveat])
    
    # If parent doesn't have this caveat, child is more restrictive
    if key not in parent_parsed:
        return True
    
    # For numeric caveats, child must be <= parent
    numeric_caveats = ['budget', 'ratelimit', 'expires']
    if any(nc in key for nc in numeric_caveats):
        try:
            parent_val = float(parent_parsed[key].split()[0])
        except (ValueError, IndexError):
            parent_val = float('inf')
        try:
            child_val = float(child_parsed[key].split()[0])
        except (ValueError, IndexError):
            child_val = 0
        return child_val <= parent_val
    
    # For scope caveats, child must be subset of parent
    if key == 'scope':
        parent_scopes = parent_parsed[key].split(',')
        child_scopes = child_parsed[key].split(',')
        return all(
            any(
                ps.strip() == cs.strip() or 
                ps.strip() == '*' or 
                (ps.strip().endswith(':*') and cs.strip().startswith(ps.strip()[:-1]))
                for ps in parent_scopes
            )
            for cs in child_scopes
        )
    
    return True  # Default to allowing
